File: mobile/ios/TigerWallet/Services/Trading/CopyTradingService.swift
//
//  CopyTradingService.swift
//  TigerWallet
//
//  Copy Trading Service - iOS Implementation
//

import Foundation

struct Trader {
    var id: String
    var address: String
    var username: String
    var avatar: String
    var winRate: Double
    var totalPnl: Double
    var followers: Int
    var copyCount: Int
    var tradingPair: String
    var monthlyPnl: Double
    var weeklyPnl: Double
    var dailyPnl: Double
    var riskLevel: String
    var isFollowing: Bool
    var isVerified: Bool
}

struct CopyPosition {
    var id: String
    var traderId: String
    var traderName: String
    var userId: String
    var symbol: String
    var side: String
    var size: Double
    var entryPrice: Double
    var currentPrice: Double
    var pnl: Double
    var pnlPercent: Double
    var openTime: Date
    var status: String
}

class CopyTradingService {
    
    private var traders: [Trader] = []
    
    init() {
        initializeTraders()
    }
    
    private func initializeTraders() {
        traders = [
            Trader(id: "trader_0", address: "0x742d35Cc6634C0532925a3b844Bc9e7595f0eB1E", username: "TraderAlex", avatar: "🐋", winRate: 78.5, totalPnl: 125000, followers: 5420, copyCount: 2710, tradingPair: "BTC/USDT", monthlyPnl: 31250, weeklyPnl: 7500, dailyPnl: 1250, riskLevel: "MEDIUM", isFollowing: false, isVerified: true),
            Trader(id: "trader_1", address: "0x1234567890abcdef1234567890abcdef12345678", username: "CryptoKing", avatar: "👑", winRate: 72.3, totalPnl: 98500, followers: 3210, copyCount: 1605, tradingPair: "ETH/USDT", monthlyPnl: 24625, weeklyPnl: 5910, dailyPnl: 985, riskLevel: "HIGH", isFollowing: true, isVerified: true),
            Trader(id: "trader_2", address: "0xabcdef1234567890abcdef1234567890abcdef12", username: "DeFiMaster", avatar: "🎯", winRate: 85.0, totalPnl: 150000, followers: 8930, copyCount: 4465, tradingPair: "SOL/USDT", monthlyPnl: 37500, weeklyPnl: 9000, dailyPnl: 1500, riskLevel: "LOW", isFollowing: false, isVerified: true),
            Trader(id: "trader_3", address: "0x9876543210fedcba9876543210fedcba98765432", username: "AltSeason", avatar: "🚀", winRate: 65.0, totalPnl: 87000, followers: 1890, copyCount: 945, tradingPair: "XRP/USDT", monthlyPnl: 21750, weeklyPnl: 5220, dailyPnl: 870, riskLevel: "HIGH", isFollowing: true, isVerified: false),
            Trader(id: "trader_4", address: "0xabcdefabcdefabcdefabcdefabcdefabcdefabcd", username: "BitcoinWhale", avatar: "🐋", winRate: 82.0, totalPnl: 200000, followers: 6540, copyCount: 3270, tradingPair: "BTC/USDT", monthlyPnl: 50000, weeklyPnl: 12000, dailyPnl: 2000, riskLevel: "MEDIUM", isFollowing: false, isVerified: true)
        ]
    }
    
    func getTopTraders(limit: Int = 10) -> [Trader] {
        return Array(traders.prefix(limit))
    }
    
    func getAllTraders() -> [Trader] {
        return traders
    }
    
    func getTrader(traderId: String) -> Trader? {
        return traders.first { $0.id == traderId }
    }
    
    func followTrader(traderId: String) {
        if let index = traders.firstIndex(where: { $0.id == traderId }) {
            traders[index].isFollowing.toggle()
            traders[index].followers += traders[index].isFollowing ? 1 : -1
        }
    }
    
    func copyTrade(userId: String, traderId: String, symbol: String, side: String, amount: Double) -> CopyPosition {
        let trader = getTrader(traderId: traderId)
        
        return CopyPosition(
            id: "copy_\(Int(Date().timeIntervalSince1970 * 1000))",
            traderId: traderId,
            traderName: trader?.username ?? "Unknown",
            userId: userId,
            symbol: symbol,
            side: side,
            size: amount,
            entryPrice: 43250.0,
            currentPrice: 43250.0,
            pnl: 0,
            pnlPercent: 0,
            openTime: Date(),
            status: "OPEN"
        )
    }
    
    func searchTraders(query: String) -> [Trader] {
        return traders.filter {
            $0.username.lowercased().contains(query.lowercased()) ||
            $0.address.lowercased().contains(query.lowercased())
        }
    }
}

File: desktop_wallet/src/services/copy_trading_service.cpp
// Copy Trading Service Implementation
#include "copy_trading_service.h"
#include <sstream>
#include <algorithm>

namespace tigerwallet {

CopyTradingService::CopyTradingService() {}

CopyTradingService::~CopyTradingService() {}

void CopyTradingService::initializeTraders() {
    // The trader list is sourced from the backend copy-trading service. There
    // is currently no such endpoint on the wallet_api backend, so we start
    // with an EMPTY list rather than seeding fabricated "demo" traders. Never
    // seed mock traders with invented addresses/PnL.
    traders_.clear();
}

std::vector<Trader> CopyTradingService::getTopTraders(int limit) {
    std::vector<Trader> sorted = traders_;
    std::sort(sorted.begin(), sorted.end(), [](const Trader& a, const Trader& b) {
        return a.totalPnL > b.totalPnL;
    });
    if (limit > 0 && limit < (int)sorted.size()) {
        sorted.resize(limit);
    }
    return sorted;
}

std::vector<Trader> CopyTradingService::searchTraders(const std::string& query) {
    std::vector<Trader> results;
    for (const auto& trader : traders_) {
        if (trader.username.find(query) != std::string::npos ||
            trader.address.find(query) != std::string::npos) {
            results.push_back(trader);
        }
    }
    return results;
}

std::vector<Trader> CopyTradingService::getFollowing(const std::string& userId) {
    if (following_.find(userId) != following_.end()) {
        return following_[userId];
    }
    return std::vector<Trader>();
}

void CopyTradingService::followTrader(const std::string& userId, const std::string& traderId) {
    for (auto& trader : traders_) {
        if (trader.id == traderId) {
            trader.followers++;
            trader.isFollowing = true;
            following_[userId].push_back(trader);
            break;
        }
    }
}

void CopyTradingService::unfollowTrader(const std::string& userId, const std::string& traderId) {
    for (auto& trader : traders_) {
        if (trader.id == traderId) {
            trader.followers--;
            trader.isFollowing = false;
            break;
        }
    }
    
    auto& userFollowing = following_[userId];
    userFollowing.erase(
        std::remove_if(userFollowing.begin(), userFollowing.end(),
            [&traderId](const Trader& t) { return t.id == traderId; }),
        userFollowing.end()
    );
}

std::vector<CopyPosition> CopyTradingService::getCopyPositions(const std::string& userId) {
    if (positions_.find(userId) != positions_.end()) {
        return positions_[userId];
    }
    return std::vector<CopyPosition>();
}

CopyPosition CopyTradingService::copyTrade(const std::string& userId, const std::string& traderId,
                                           const std::string& symbol, const std::string& side, double amount) {
    CopyPosition pos;
    std::stringstream ss;
    ss << "copy_" << time(nullptr);
    pos.id = ss.str();
    pos.traderId = traderId;
    pos.userId = userId;
    pos.symbol = symbol;
    pos.side = side;
    pos.size = amount;
    pos.entryPrice = 0.0; // No fabricated price; sourced from the real market via the backend
    pos.currentPrice = 0.0;
    pos.pnl = 0;
    pos.pnlPercent = 0;
    pos.openTime = time(nullptr);
    pos.status = "OPEN";
    
    // Find trader name
    for (const auto& trader : traders_) {
        if (trader.id == traderId) {
            pos.traderName = trader.username;
            break;
        }
    }
    
    positions_[userId].push_back(pos);
    return pos;
}

CopySettings CopyTradingService::getSettings(const std::string& userId) {
    if (settings_.find(userId) != settings_.end()) {
        return settings_[userId];
    }
    CopySettings defaultSettings;
    defaultSettings.userId = userId;
    settings_[userId] = defaultSettings;
    return defaultSettings;
}

void CopyTradingService::updateSettings(const std::string& userId, const CopySettings& settings) {
    settings_[userId] = settings;
}

void CopyTradingService::closeCopyPosition(const std::string& userId, const std::string& positionId) {
    auto& userPositions = positions_[userId];
    for (auto& pos : userPositions) {
        if (pos.id == positionId) {
            pos.status = "CLOSED";
            // Closing PnL must be computed from the real exit price sourced
            // from the backend/market. Without it we leave the PnL fields
            // unchanged (honest "pending") rather than inventing a random
            // outcome. Final settlement is performed by the backend.
            break;
        }
    }
}

} // namespace tigerwallet
